Fix race condition where concurrently loading a namespaced class results in LoadError

DimitriosLisenko opened the pull request:
### Summary

Full description in ticket #33209 - in short, it looks like there is a race condition when concurrently loading namespaced classes. Each thread will potentially create its own local copy of the module and only one of the threads will define the class on it, causing the others to fail with `LoadError` since their local module objects will not have the class defined on them.

Copy of #33210 against a different base branch - @matthewd I'll address your comments in that PR once I thought about it some more :)

DimitriosLisenko committed:
Fix race condition where concurrently loading a namespaced class results in LoadError

javan commented:
I'm pretty sure this a bootsnap issue that's fixed, but unreleased. See https://github.com/rails/rails/issues/33209#issuecomment-399774731. Please do correct me if I'm wrong.

javan commented:
Reopened. See https://github.com/rails/rails/issues/33209#issuecomment-399775297, @matthewd.

DimitriosLisenko committed:
Add executable test case for autoload_module! race condition

DimitriosLisenko commented:
Hey @matthewd - addressing your issues in https://github.com/rails/rails/pull/33210#issuecomment-399757361.

> Yes please.

Made this PR against master.

> Have you attempted to write a test for this? If so, where did you get stuck?

Added a test now (https://github.com/rails/rails/pull/33211/commits/12f8283732e217cbdc2394597126d8c55c4dafad) - I was a bit concerned I'd have to implement channels to properly replicate the timing issue, but it looks like just calling `const_missing` twice on the main thread is sufficient.

> Should/could this use the load interlock instead of a separate mutex?

I hadn't considered that - I was under the impression the load interlock should be used for loading files. It makes sense though! That also means the `autoloaded_constants` will only get edited under the load/unload interlock now, which is good.

An issue I can see is this: what happens if `autoload_module!` (which is under the load interlock) is called in order to load a module into another module (e.g. for the class `Deep::Really::Super::Deep`, we're at the point where `Super` is being loaded into `Deep::Really`) but directly before that, `ActiveSupport::Dependencies.clear` (which is under the unload interlock) is run - this will result in an invalid state for `autoloaded_constants`. This is also an issue for `require_or_load` (which is already under the load interlock).

I'm thinking we can follow this line https://github.com/rails/rails/blob/df20bf4aa25d9d7680bba3448150acde3a4e0ad1/activesupport/lib/active_support/dependencies.rb#L450
with
```ruby
return nil unless Object.const_defined?(into.name, false)
````
and then let that case be handled by
https://github.com/rails/rails/blob/df20bf4aa25d9d7680bba3448150acde3a4e0ad1/activesupport/lib/active_support/dependencies.rb#L538
though for a deeply nested module, it will result in https://github.com/rails/rails/blob/df20bf4aa25d9d7680bba3448150acde3a4e0ad1/activesupport/lib/active_support/dependencies.rb#L491
being thrown as the code currently is - maybe `const_missing` should be called recursively? Anyway, I'll give it some proper thought once I can find a chunk of time, just wondering if you have any thoughts on this.

> Given this lock, do we still need the `uniq!`?

Agree - don't think it's still necessary.

DimitriosLisenko committed:
Use the load interlock for autoloading modules

shepmaster commented:
I'd love to see this fix get backported to the 5.0 branch; we are in the process of upgrading past that but it might be a while ;-)

oniofchaos commented:
Unless there's any other issues to address based on the other comments, this seems 👍 after addressing the merge conflict.

DimitriosLisenko commented:
There are some issues to address still, but unfortunately I haven't had the time to look into it yet. I wouldn't consider this PR mergeable as it stands.

DimitriosLisenko commented:
Not in any practical way unfortunately - I've been the version of `const_missing` in https://github.com/rails/rails/issues/33209#issuecomment-429283496 for a while now locally, and while it stops some class of errors, I still get NameErrors fairly regularly when running with multiple threads.

shepmaster commented:
> these issues in your tests

Oh we have, thank you. However, we still see it in development because we have a JS frontend that makes multiple parallel requests. Every time we switch branches on these pages, we hit the error. Reloading a second time works, but is extremely annoying.
